[Merged by Bors] - feat(*): define classes for coercions that are homomorphisms

Vierkantor opened the pull request:
This PR is part of my [plan to refactor `algebra_map`](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Should.20.60algebra_map.60.20be.20a.20coercion.3F/near/297206409). It introduces new classes that should connect coercions (`↑`s) and bundled homomorphisms, e.g. `ring_hom`, so that e.g. an instance of `coe_ring_hom R S` states that there is a canonical ring homomorphism `ring_hom.coe R S` from `R` to `S`. These classes are unbundled (they take an instance of `has_lift_t R S` as a parameter, rather than extending `has_lift_t` or one of its subclasses) for two reasons:

 * We wouldn't have to introduce new classes that handle transitivity (and probably cause diamonds)
 * It doesn't matter whether a coercion is written with `has_coe` or `has_lift`, you can give it a homomorphism structure in exactly the same way.

# Additions

 * classes `coe_add_hom M N`, `coe_mul_hom`, `coe_one_hom`, `coe_zero_hom` state that the coercion map `coe : M → N` preserves the corresponding operation. They take in a `has_lift_t` instance so it doesn't matter how precisely the coercion map is defined as long as Lean can find the instance.
 * classes `coe_monoid_hom M N`, `coe_add_monoid_hom M N`, `coe_monoid_with_zero_hom M N`, `coe_non_unital_ring_hom R S` and `coe_ring_hom R S` are convenient groupings of the above classes 
 * simp lemmas `coe_add`, `coe_mul`, `coe_zero`, `coe_one`: basic lemmas stating that `coe` preserves the corresponding operation.
 * simp lemmas `coe_pow`, `coe_nsmul`, `coe_zpow`, `coe_zsmul`, `coe_bit0`, `coe_bit1`, `coe_sub`, `coe_neg`:  lemmas stating `coe` preseves the derived structure
 * simp lemmas `coe_inv`, `coe_div` (for groups), `coe_inv₀`, `coe_div₀` (for groups with zero). An alternative to having this pair of otherwise identical lemmas would be to add `coe_inv_hom` and `coe_div_hom` classes.
 * class `coe_smul_hom M X Y` stating that `coe : X → Y` preseves scalar multiplication by elements of `M`
 * class `coe_semilinear_map σ M N` stating that `coe : M → N` is `σ`-semilinear. `coe_linear_map R M N` is a synonym of `coe_semilinear_map (ring_hom.id R) M N` and also gives an instance of `coe_smul_hom`. I don't think this has any applications yet, it was mostly to test that this design would also work for more complicated classes.
 * definitions `mul_hom.coe`, `add_hom.coe`, `one_hom.coe`, `zero_hom.coe`, `monoid_hom.coe`, `add_monoid_hom.coe`, `monoid_with_zero_hom.coe`, `non_unital_ring_hom.coe`, `ring_hom.coe`, `semilinear_map.coe` and `linear_map.coe`, given the appropriate `coe_hom` class, are bundled forms of the coercion map that correspond to the `coe_hom` classes. In particular `ring_hom.coe` should become an alternative to `algebra_map`.
 * `mul_action_hom.coe`, `distrib_mul_action_hom.coe` and `mul_semiring_action_hom.coe` are bundled forms of the coercion map corresponding to `coe_smul_hom` (since they only vary in hypotheses on the non-bundled structure)

# Changes

Whenever existing lemmas had a name conflict with the new generic `coe_` lemmas, I have made them `protected`. In many places, the new lemmas are silently used instead of the now `protected` old lemmas. This caused the vast majority of changed lines. The full list is:

 * `add_monoid_hom.coe_mul` (is about `coe_fn`)
 * `linear_map.coe_one` and `linear_map.coe_mul` (are about `coe_fn` and endomorphisms)
 * `submodule.coe_zero`, `submodule.coe_add`, `submodule.coe_smul` (superseded)
 * `dfinsupp.coe_zero`, `dfinsupp.coe_add`, `dfinsupp.coe_nsmul`, `dfinsupp.coe_neg`, `dfinsupp.coe_sub`, `dfinsupp.coe_zsmul`, `dfinsupp.coe_smul` (are about `coe_fn`)
 * `finset.coe_one` (superseded)
 * `finsupp.coe_smul`, `finsupp.coe_zero`, `finsupp.coe_add`, `finsupp.coe_neg`, `finsupp.coe_sub`, `finsupp.coe_mul` (are about `coe_fn`)
 * `enat.coe_zero`, `enat.coe_one`, `enat.coe_add`, `enat.coe_sub`, `enat.coe_mul` (superseded except `coe_sub`)
 * `nnrat.coe_zero`, `nnrat.coe_one`, `nnrat.coe_add`, `nnrat.coe_mul`, `nnrat.coe_inv`, `nnrat.coe_div`, `nnrat.coe_bit0`, `nnrat.coe_bit1`, `nnrat.coe_sub` (superseded except `coe_sub`)
 * `ereal.coe_zero`, `ereal.coe_one`, `ereal.coe_add`, `ereal.coe_mul`, `ereal.coe_nsmul`, `ereal.coe_pow`, `ereal.coe_bit0`, `ereal.coe_bit1` (superseded)
 * `ennreal.coe_zero`, `ennreal.coe_one`, `ennreal.coe_add`, `ennreal.coe_mul`, `ennreal.coe_bit0`, `ennreal.coe_bit1`, `ennreal.coe_pow`, `ennreal.coe_inv`, `ennreal.coe_div` (superseded)
 * `equiv.perm.coe_one` and `equiv.perm.coe_mul` (are about `coe_fn` and endomorphisms)
 * `submonoid.coe_one`, `submonoid.coe_mul`, `submonoid.coe_pow` (superseded)
 * `subsemigroup.coe_mul` (superseded)
 * `special_linear_group.coe_inv`, `special_linear_group.coe_mul`, `special_linear_group.coe_pow`
 * `vector_measure.coe_smul`, `vector_measure.coe_zero`, `vector_measure.coe_add`, `vector_measure.coe_neg` (are about `coe_fn`)
 * `lucas_lehmer.X.coe_mul` (superseded)
 * `power_series.coe_add`, `power_series.coe_zero`, `power_series.coe_mul`, `power_series.coe_one`, `power_series.coe_bit0`, `power_series.coe_bit1`, `power_series.coe_pow` (superseded)
 * `uniform_space.completion.coe_zero`, `uniform_space.completion.coe_neg`, `uniform_space.completion.coe_sub`, `uniform_space.completion.coe_add` (superseded)
 * `continuous_function.coe_mul`, `continuous_function.coe_one`, `continuous_function.coe_inv`, `continuous_function.coe_div` `continuous_function.coe_pow`, `continuous_function.coe_zpow` (are about `coe_fn)
 * `bounded_continuous_function.coe_zero`, `bounded_continuous_function.coe_one`, `bounded_continuous_function.coe_add`, `bounded_continuous_function.coe_neg`, `bounded_continuous_function.coe_sub`, `bounded_continuous_function.coe_mul`, `bounded_continuous_function.coe_nsmul`, `bounded_continuous_function.coe_zsmul`, `bounded_continuous_function.coe_pow`,

Many `coe_` or `cast_` lemmas have been reimplemented using `coe_hom` classes and are no longer `@[simp]`. Many homomorphisms `subfoo.subtype` have been reimplemented by `foo_hom.coe`. We might consider replacing all of these outright.

Finally, some `simp` calls had to be fixed, e.g.  `simp [← something.cast_add]` could loop due to the new `coe_add` lemma, so it had to become `simp [← something.cast_add, -coe_add]`.

# Further plans

Some obvious further steps that I didn't feel fit in the (already wide) scope of this PR:

 * Reviewing for missing instances: currently I only added the obvious instances for the new classes, so we should go through and ensure everything has its instance.
 * Reviewing for missing `coe` lemmas: currently I only added lemmas that existed in the `submonoid`/`subgroup`/`subring` files, so we should go through and ensure everything has its lemma.
 * Instances for the transitivity instances `coe_trans` and `lift_trans`, so if `coe : M → N` is a monoid hom and `coe : N → O` is a monoid hom then `coe : M → O` is also a monoid hom.
 * Classes for when `coe_fn` is a homomorphism: many of the naming conflicts that required `protected` are of the form `mul_hom.coe_mul : ⇑(f * g) =  ⇑f * ⇑g`, so the obvious consequence is adding a class `coe_fn_mul_hom` along these same lines as in this PR.
 * Combining with the `algebra.to_has_lift_t` instance: this is likely to require some subtle fixing so I'd rather merge the big changes first so I don't have to keep those up to date as well.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor committed:
Add `coe_ring_hom`

Vierkantor committed:
Add `coe_inv` and `coe_div` lemmas

Vierkantor committed:
Instances for `ℕ` and `ℤ`

Vierkantor committed:
`protected` against confusion

Vierkantor committed:
`protected` defintions of `coe`

Vierkantor committed:
Redefine `sub{semigroup,monoid,group}.subtype` using the `coe` classes

Vierkantor committed:
`coe_pow` lemmas

Vierkantor committed:
`coe_ring_hom` for `ℚ`

Vierkantor committed:
Use the coe classes for sub(semi)rings

Vierkantor committed:
Add `norm_cast` attributes

Vierkantor committed:
`coe_ring_hom` instance for `enat`

Vierkantor committed:
Fix this simp call looping between `coe_add` and `←nat.cast_add`

Vierkantor committed:
`coe_bit0` and `coe_bit1` lemmas

Vierkantor committed:
Remove `simp`/`norm_cast` attributes from `nat` and `int` cast

Vierkantor committed:
Re-add `@[norm_cast]` attribute (not completely sure why it's needed)

Vierkantor committed:
More `cast` lemmas that don't need `@[simp]` anymore

Vierkantor committed:
`coe_ring_hom` should also weaken to `coe_monoid_with_zero_hom`

Vierkantor committed:
Add special-case `coe_one_hom` instance for `nat`

Apparently ordinals are not enough of a ring to warrant a `coe_ring_hom` instance!

Vierkantor committed:
Add classes for coercions that are linear/semilinear maps

Vierkantor committed:
`coe_hom` instance for submodule → module coercion

Vierkantor committed:
Make some `finsupp.coe_` lemmas `protected`

Vierkantor committed:
Protected `dfinsupp.coe_` lemmas

Vierkantor committed:
Fix `data/zmod/basic` (`simp` doesn't pick up on `coe_zero` but does like `int.cast_zero`)

Vierkantor committed:
Fix `mul_comm` firing at a different moment while `simp`ing

Vierkantor committed:
Protect `coe_` lemmas for `uniform_space.completion`

Vierkantor committed:
More `protected`

Vierkantor committed:
More `protected` in `finsupp` and `dfinsupp`

Vierkantor committed:
Make `continuous_map.coe_` lemmas protected

Vierkantor committed:
Make `add_monoid_hom.coe_mul` protected

Vierkantor committed:
`coe_ring_hom` instance for `power_series` (+ protected)

Vierkantor committed:
Make `lucas_lehmer.X.coe_mul` `protected`

Vierkantor committed:
More `protected`

Vierkantor committed:
Even more `protected`

Vierkantor committed:
Fix missing names (apparently there's now a name conflict with `power_series`?)

Vierkantor committed:
`protected` in `linear_map.coe_` and `special_linear_group.coe_`

Vierkantor committed:
Fix (seems to be `simp` order messing things up?)

Vierkantor committed:
Fix more `protected`

Vierkantor committed:
Fix `simp` loop due to new lemmas

Vierkantor committed:
Fix (changed `simp` set)

Vierkantor committed:
Fix (protected)

Vierkantor committed:
Make `vector_measure.coe_` lemmas protected

Vierkantor committed:
Fix `simp` set due to `protected`

Vierkantor committed:
Fix `analysis/inner_product_space/dual.lean` (protected)

Vierkantor committed:
Protected `finset.coe_one`

Vierkantor committed:
Protected `ennreal.coe_` lemmas

Vierkantor committed:
Fix `simp` loop due to `←` + new lemmas

Vierkantor committed:
Fix: protected

Vierkantor committed:
Fix that `measure.coe_add` is protected

Vierkantor committed:
Fix (`cast_add` has new synonym `coe_add`)

Vierkantor committed:
Fix unprotected usage of `ennreal.coe_inv`

Vierkantor committed:
Lint fix

Vierkantor committed:
Fix: `complex.coe_smul` should be protected

Vierkantor committed:
Fix test: `ennreal.coe_div` is protected

Vierkantor committed:
Lint fixing

Vierkantor committed:
Fix parameter count

Vierkantor committed:
Make `ereal.coe_` lemmas protected; fix `simp` looping when rewriting to the left with specific lemmas

Vierkantor committed:
Oops, missed a loop

Vierkantor committed:
Fix unprotected reference to `continuous_map.coe_`

Vierkantor committed:
Do the same but actually correctly this time

Vierkantor committed:
Fix more protected `continuous_map.coe_`

Vierkantor committed:
Fix protected `complex.coe_smul`

Vierkantor committed:
Fix protected `finsupp.coe_add`

Vierkantor committed:
Lint fix: `to_additive` used in the wrong order, so it wasn't applying nolint

Vierkantor committed:
Fix usage of `dfinsupp.coe_zero`

eric-wieser commented on src/algebra/hom/group.lean:
```diff
@@ -1236,3 +1236,140 @@ protected lemma commute.map [mul_hom_class F M N] (h : commute x y) (f : F) :
+
+/-! ### Coercions as bundled morphisms
+
+The classes `coe_mul_hom`, `coe_monoid_hom`, etc. state that the coercion map `↑` (a.k.a. `coe`)
```
Can you copy the explanation from the PR for why you use `has_lift_t` and not `has_coe` to this docstring?

Vierkantor commented on src/algebra/hom/group.lean:
```diff
@@ -1236,3 +1236,140 @@ protected lemma commute.map [mul_hom_class F M N] (h : commute x y) (f : F) :
+/-! ### Coercions as bundled morphisms
+
+The classes `coe_mul_hom`, `coe_monoid_hom`, etc. state that the coercion map `↑` (a.k.a. `coe`)
+is a homomorphism.
```
```suggestion
The classes `coe_mul_hom`, `coe_monoid_hom`, etc. state that the coercion map `↑` (a.k.a. `coe`)
is a homomorphism.
These classes are unbundled (they take an instance of `has_lift_t R S` as a parameter, rather than
extending `has_lift_t` or one of its subclasses) for two reasons:

 * We wouldn't have to introduce new classes that handle transitivity (and probably cause diamonds)
 * It doesn't matter whether a coercion is written with `has_coe` or `has_lift`, you can give it
   a homomorphism structure in exactly the same way.
```

Vierkantor committed:
Rename `coe_foo_hom` to `coe_is_foo_hom`

Vierkantor committed:
Comment on the inheritance structure of these classes

Vierkantor committed:
Linx fix: overlong lines

Vierkantor committed:
Also rename the `to_is_foo_hom` instances

Vierkantor committed:
Fix missing `_is_`

Vierkantor committed:
`coe_ring_hom` for `nnrat`

eric-wieser commented on src/algebra/hom/group.lean:
```diff
@@ -1237,3 +1237,147 @@ protected lemma commute.map [mul_hom_class F M N] (h : commute x y) (f : F) :
+
+section coe
+
+/-! ### Coercions as bundled morphisms
```
Can you mention the things in this section in the module docstring?

eric-wieser commented on src/algebra/group_power/lemmas.lean:
```diff
@@ -358,14 +358,14 @@ instance non_unital_non_assoc_semiring.nat_is_scalar_tower [non_unital_non_assoc
 
-@[simp, norm_cast] theorem nat.cast_pow [semiring R] (n m : ℕ) : (↑(n ^ m) : R) = ↑n ^ m :=
+@[norm_cast] theorem nat.cast_pow [semiring R] (n m : ℕ) : (↑(n ^ m) : R) = ↑n ^ m :=
 begin
   induction m with m ih,
   { rw [pow_zero, pow_zero], exact nat.cast_one },
   { rw [pow_succ', pow_succ', nat.cast_mul, ih] }
 end
```
Can this be proved with the new lemmas, or are they not available yet here?

Vierkantor replied:
The instance isn't defined yet here, unfortunately.

eric-wieser commented on src/algebra/module/linear_map.lean:
```diff
@@ -951,3 +951,48 @@ lemma End.int_cast_def (z : ℤ) [add_comm_group N₁] [module R N₁] :
+variables {R S} [semiring R] [semiring S] (σ : R →+* S)
+
+/-- `coe_is_semilinear_map σ M N` is a class stating that the coercion map `↑ : M → N`
+(a.k.a. `coe`) is a `σ`-semilinear map.
```
Perhaps something like "see the notes for `coe_is_smul_hom` for an explanation of why `coe_is_add_monoid_hom` must be provided separately"; or just repeat that note here.

Vierkantor replied:
I went with a new comment since the situation isn't exactly the same.

eric-wieser commented on src/algebra/module/submodule/basic.lean:
```diff
@@ -208,12 +208,19 @@ protected lemma nonempty : (p : set M).nonempty := ⟨0, p.zero_mem⟩
 
+instance : coe_is_add_monoid_hom p M :=
+{ coe_zero := rfl,
+  coe_add := λ _ _, rfl }
```
Does this generalize to all `add_submonoid_class`es?

Vierkantor replied:
Indeed, that instance already exists so we don't need this instance at all!

eric-wieser commented on src/algebra/module/submodule/basic.lean:
```diff
@@ -208,12 +208,19 @@ protected lemma nonempty : (p : set M).nonempty := ⟨0, p.zero_mem⟩
-@[norm_cast] lemma coe_smul (r : R) (x : p) : ((r • x : p) : M) = r • ↑x := rfl
+@[norm_cast] protected lemma coe_add (x y : p) : (↑(x + y) : M) = ↑x + ↑y := rfl
+@[norm_cast] protected lemma coe_zero : ((0 : p) : M) = 0 := rfl
+@[norm_cast] protected lemma coe_smul (r : R) (x : p) : ((r • x : p) : M) = r • ↑x := rfl
```
I'd argue these are worth keeping as `simp` because they're true by `rfl` and therefore elegible for `dsimp`.

eric-wieser commented on src/analysis/special_functions/pow.lean:
```diff
@@ -1606,7 +1606,7 @@ begin
 
 @[simp] lemma one_rpow (x : ℝ) : (1 : ℝ≥0∞) ^ x = 1 :=
-by { rw [← coe_one, coe_rpow_of_ne_zero one_ne_zero], simp }
+by { rw [← @coe_one ℝ≥0, coe_rpow_of_ne_zero one_ne_zero], simp }
```
I think this is a good argument that `coe_one` and `coe_zero` should take an explicit type argument

Vierkantor replied:
Is it possible to specify in structure fields that the first parameter is explicit and the second is implicit? `(coe_zero [] : _)` makes both explicit.

eric-wieser replied:
I suspect you have to write manual aliases wrapping the projection.

At any rate, this can wait till a follow-up

eric-wieser commented on src/data/enat/basic.lean:
```diff
@@ -30,11 +30,17 @@ instance : is_well_order ℕ∞ (<) := { }
+protected lemma coe_one : ((1 : ℕ) : ℕ∞) = 1 := rfl
+protected lemma coe_add (m n : ℕ) : ↑(m + n) = (m + n : ℕ∞) := rfl
+protected lemma coe_sub (m n : ℕ) : ↑(m - n) = (m - n : ℕ∞) := rfl
+protected lemma coe_mul (m n : ℕ) : ↑(m * n) = (m * n : ℕ∞) := with_top.coe_mul
```
Same comment as before about `dsimp` / `@[simp]`.

eric-wieser commented on src/data/rat/nnrat.lean:
```diff
@@ -67,15 +67,21 @@ open _root_.rat (to_nnrat)
+  coe_add := λ _ _, rfl,
+  coe_mul := λ _ _, rfl }
+
+@[simp, norm_cast] protected lemma coe_zero : ((0 : ℚ≥0) : ℚ) = 0 := rfl
+@[simp, norm_cast] protected lemma coe_one  : ((1 : ℚ≥0) : ℚ) = 1 := rfl
+@[simp, norm_cast] protected lemma coe_add (p q : ℚ≥0) : ((p + q : ℚ≥0) : ℚ) = p + q := rfl
+@[simp, norm_cast] protected lemma coe_mul (p q : ℚ≥0) : ((p * q : ℚ≥0) : ℚ) = p * q := rfl
+@[simp, norm_cast] protected lemma coe_inv (q : ℚ≥0) : ((q⁻¹ : ℚ≥0) : ℚ) = q⁻¹ := rfl
+@[simp, norm_cast] protected lemma coe_div (p q : ℚ≥0) : ((p / q : ℚ≥0) : ℚ) = p / q := rfl
+@[simp, norm_cast] protected lemma coe_bit0 (q : ℚ≥0) : ((bit0 q : ℚ≥0) : ℚ) = bit0 q := rfl
+@[simp, norm_cast] protected lemma coe_bit1 (q : ℚ≥0) : ((bit1 q : ℚ≥0) : ℚ) = bit1 q := rfl
+@[simp, norm_cast] protected lemma coe_sub (h : q ≤ p) : ((p - q : ℚ≥0) : ℚ) = p - q :=
```
Looks like you were able to keep the dsimp lemmas here, which is great!

eric-wieser commented on src/data/zmod/basic.lean:
```diff
@@ -128,7 +128,10 @@ def cast : Π {n : ℕ}, zmod n → R
+-- `zmod 0` unifies with `ℤ` but doesn't inherit the `coe_is_add_monoid_hom` instance
+-- so we add `int.cast_zero` to the `simp` set despite being an instance of the
+-- existing `@[simp] lemma coe_zero`
+by cases n; simp [int.cast_zero]
```
I don't fully understand the explanation here; can't we add a `coe_is_add_monoid_hom` instance for `zmod 0` by copying the int one?

eric-wieser commented on src/number_theory/bernoulli_polynomials.lean:
```diff
@@ -190,7 +190,7 @@ begin
 end
 
-open power_series
+open _root_.power_series
```
Accidental?

Vierkantor replied:
It was an autogenerated instance name creating a new namespace `polynomial.power_series`.

eric-wieser commented on src/ring_theory/subsemiring/basic.lean:
```diff
@@ -100,14 +103,9 @@ instance to_semiring {R} [semiring R] [set_like S R] [subsemiring_class S R] : s
-end
+@[norm_cast] lemma coe_pow {R} [semiring R] [set_like S R] [subsemiring_class S R]
+  (x : s) (n : ℕ) : ((x^n : s) : R) = (x^n : R) :=
+by apply coe_pow
```
Does `coe_pow _ _` (maybe with more `_`s) not work here?

eric-wieser reviewed (approved):
Some small comments, but generally this looks great!

Vierkantor committed:
This instance already exists for `add_submodule_class`

Vierkantor committed:
These can stay `simp` lemmas

Vierkantor committed:
Keep these `dsimp` lemmas too

Vierkantor committed:
We can indeed just add a `coe_is_hom` instance for `zmod 0`

Vierkantor committed:
More `dsimp`-eligible lemmas

Vierkantor committed:
Add `coe_is_hom` classes to the module docstring

Vierkantor committed:
Fix parameters to `coe_pow`

Vierkantor committed:
Explain why `coe_is_semilinear_map` doesn't talk about the additive structure

Vierkantor committed:
Fix that autogenerated instance name created a `polynomial.power_series` namespace

Vierkantor committed:
Protect `bounded_continuous_function.coe_` lemmas

Also includes some fixes for `nat.cast_zero` being un`@[simp]`ed and
`coe_is_zero_hom` not being available yet

eric-wieser commented on src/algebra/hom/group.lean:
```diff
@@ -28,6 +28,15 @@ building blocks for other homomorphisms:
+ * `coe_is_mul_hom`/`coe_is_add_monoid_hom`
+ * `coe_is_monoid_hom`/`coe_is_add_monoid_hom`
+ * `coe_is_monoid_with_zero_hom`
+These come with a selection of `simp` lemmas stating that `↑` preseves the corresponding operation:
```
```suggestion
These come with a selection of `simp` lemmas stating that `↑` preserves the corresponding operation:
```

Vierkantor committed:
Fix parameter usage in `ereal.lean`

Vierkantor committed:
Typo fix

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Vierkantor committed:
Fix `int.cast_zero` being un`@[simp]`ed

Vierkantor committed:
Fix `ereal.coe_add` being a (`d`)`simp` lemma so not being excluded

Vierkantor committed:
Lint fix: don't need `submodule.coe_smul` to be simp or even dsimp

Vierkantor committed:
WIP feat(algebra/hom): classes for when the coercion map `↑` is a homomorphism
